### DISCUSSION TITLE
none value will be ignored when update a associate list

### DISCUSSION THREAD

**Discussion by htyangya** [sqlalchemy/sqlalchemy#12071] [Usage Questions] [answered]
i hove two model like that:
```
class Parent(db.Model):
  ...
  children = db.relationship(Child,  cascade="all,delete-orphan")

class Child(db.Model):
  a = db.Column(db.Integer(), primary_key=True, , db.ForeignKey('Parent.id'))
  b = db.Column(db.String())
  c = db.Column(db.String())
```

-----------------------------------------
```
parent = Parent.query.get(xxx)
parent.children = [Child(a=123, b="b", c=None), ...]
db.session.add(parent)
db.session.flush()
```

If the database already has a child record with a=123， The above code will trigger an update statement like this "update child set b='b' where a= 123"

The problem is that it can't update the value of c to null, c=None will be ignored.

i use flask-sqlalchemy 2.4.0 and sqlalchemy 1.3.8.

how can i update it to null? 

in addition, if the database has no record with a=123, the code will trigger a insert sql, the none value will be successfully inserted.

**Comment by zzzeek** [comment]
> If the database already has a child record with a=123， The above code will trigger an update statement like this "update child set b='b' where a= 123"

that's definitely not what that code would do, you are creating a brand new Child object which would result in a pending object that becomes an INSERT.    You would need to merge that Child object first for that to be an UPDATE.

Overall, if you want to UPDATE a row in child, you would not be touching `parent.children` at all, just locate the existing `Child` object you are interested in and set `c=None`.

**Comment by gordthompson** [comment, marked as answer]
I am unable to reproduce your issue with vanilla SQLAlchemy 1.3.8

```python
# https://stackoverflow.com/q/79172355/2144390
# https://github.com/sqlalchemy/sqlalchemy/discussions/12071
import sqlalchemy as sa
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import Session, relationship

Base = declarative_base()

class Child(Base):
    __tablename__ = "Child"
    a = sa.Column(sa.Integer(), sa.ForeignKey("Parent.id"), primary_key=True)
    b = sa.Column(sa.String())
    c = sa.Column(sa.String())

class Parent(Base):
    __tablename__ = "Parent"
    id = sa.Column(sa.Integer, primary_key=True)
    children = relationship(Child, cascade="all,delete-orphan")


e = sa.create_engine("sqlite:///db.sqlite", echo=True)
db_session = Session(e)
parent = db_session.query(Parent).get(123)
"""
2024-11-09 13:02:53,226 INFO sqlalchemy.engine.base.Engine SELECT CAST('test plain returns' AS VARCHAR(60)) AS anon_1
2024-11-09 13:02:53,226 INFO sqlalchemy.engine.base.Engine ()
2024-11-09 13:02:53,229 INFO sqlalchemy.engine.base.Engine SELECT CAST('test unicode returns' AS VARCHAR(60)) AS anon_1
2024-11-09 13:02:53,229 INFO sqlalchemy.engine.base.Engine ()
2024-11-09 13:02:53,231 INFO sqlalchemy.engine.base.Engine BEGIN (implicit)
2024-11-09 13:02:53,262 INFO sqlalchemy.engine.base.Engine SELECT "Parent".id AS "Parent_id" 
FROM "Parent" 
WHERE "Parent".id = ?
2024-11-09 13:02:53,262 INFO sqlalchemy.engine.base.Engine (123,)
"""
parent.children = [Child(a=123, b="b", c=None),]
"""
2024-11-09 13:04:24,252 INFO sqlalchemy.engine.base.Engine SELECT "Child".a AS "Child_a", "Child".b AS "Child_b", "Child".c AS "Child_c" 
FROM "Child" 
WHERE ? = "Child".a
2024-11-09 13:04:24,252 INFO sqlalchemy.engine.base.Engine (123,)
"""
db_session.flush()
"""
2024-11-09 13:05:43,271 INFO sqlalchemy.engine.base.Engine UPDATE "Child" SET b=?, c=? WHERE "Child".a = ?
2024-11-09 13:05:43,271 INFO sqlalchemy.engine.base.Engine ('b', None, 123)
"""
```

If you really are seeing the behaviour you describe then it may be a flask-sqlalchemy issue.

**Reply by htyangya** [reply]
Sorry, I did a minimal demo for test, and I did not reproduce the same problem, but it does exist in the actual program, so it is probably caused by other parts of the program. I will do another experiment to confirm where the problem lies before asking questions. Thank you.

**Reply by htyangya** [reply]
i found the answer: if i Explicitly define the value of "c", like this, it will update.

`parent.children = [Child(a=123, b="b", c=None), ...]`

but if i just use the defualt none value of "c", it will not update, the original value will remain unchanged.

`parent.children = [Child(a=123, b="b"), ...]`

**Reply by CaselIT** [reply]
that's the expected behavior. The default is "no value", that's different from None
